### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by Salamehar**
Key improvements in the refactored code:

- Enhanced Structure (Grouped related enums into nested classes, added EnumBase for consistent enum handling improved class organisation and hierarchy)

- Better Type Safety (Using Decimal for monetary values, added more comprehensive validators, made models immutable)

- Unified Rate Model (Combined all rate types into a single Rate class, added rate_type field for differentiation, maintained all necessary fields for different rate types, added validation for rate-specific fields)

- Improved Validation (Added cross-field validation, time sequence validation, rate type consistency checks)

Reasons for unified Rate model:
* Simpler database schema
* Easier querying
* More flexible for future changes
* Better relationship management
* Reduced code duplication
* Easier maintenance

The model immutability benefits:
* Immutability ensures that once a Tariff object is created, its attributes remain consistent and cannot be accidentally modified elsewhere in the code.
* Immutable objects are inherently thread-safe because their state cannot change, which is especially useful in concurrent or multi-threaded applications.
* Encourages a functional approach, where data is immutable, and modifications are performed by creating new instances instead of mutating existing ones.
* Immutable models ensure that any constraints or validation applied during object creation remain enforced, as attributes cannot be directly modified to bypass validation logic.

Example:
In the Tariff model, attributes like dno_region, rate_type, and rates are critical to maintaining the correctness of the data. If these values are accidentally modified post-creation (e.g., changing rate_type without updating rates accordingly), it could lead to logical inconsistencies. Immutabillity eliminates this risk.

### CODE REVIEW

File: opentariff/models/models.py [hunk 1]
```diff
@@ -1,113 +1,232 @@
...
+    class PaymentMethod(str, EnumBase):
+        DIRECT_DEBIT = "direct_debit"
+        PREPAYMENT = "prepayment"
+        CASH_CHEQUE = "cash_cheque"
+    
+    class TCRBandType(str, EnumBase):
+        line_loss = "line_loss"
+        consumption = "consumption"
+
+class TCRBAND_SUGGESTION(BaseModel):
```
Comment (ClemAtt) [inline opentariff/models/models.py:71]: If this class determines the standing charge, do you think it should be named `StandingCharge`? This would match calling the other classes `...Rate`.

File: opentariff/models/models.py [hunk 2]
```diff
@@ -1,113 +1,232 @@
...
+    tcrbandtype: TariffEnums.TCRBandType
+    max_consumption: Decimal
+    min_consumption: Decimal
+    line_loss: Decimal
```
Comment (ClemAtt) [inline opentariff/models/models.py:75]: @DangerousDyl is the line loss determination of the standing charge on a 1-2-1 basis or on a range?

File: opentariff/models/models.py [hunk 3]
```diff
@@ -1,113 +1,225 @@
...
+    """Base class for enums to ensure consistent string representation"""
+
+    @classmethod
+    def values(cls) -> list:
+        return [member.value for member in cls]
+
+
+class DayOfWeek(int, EnumBase):
+    ALL = 0
     MONDAY = 1
```
Comment (ClemAtt) [inline opentariff/models/models.py:17]: MONDAY should be zero
SUNDAY should be 6

This matches python default day indexing

We don't need all. It can just be None

File: opentariff/models/models.py [hunk 4]
```diff
@@ -1,113 +1,225 @@
...
+    model_config = ConfigDict(frozen=True)
+
+    rate_type: TariffEnums.RateType
+    unit_rate: Decimal = Field(..., gt=0, lt=100)
+
+
+    # Fields for time-of-use static rates
+    time_from: Optional[time] = None
+    time_to: Optional[time] = None
+    day_of_week: Optional[DayOfWeek] = None
```
Comment (ClemAtt) [inline opentariff/models/models.py:137]: Change to `day_from` and `day_to`

File: opentariff/models/models.py [hunk 5]
```diff
@@ -1,113 +1,225 @@
...
+    month_from: Optional[int] = Field(None, ge=1, le=12)
+    month_to: Optional[int] = Field(None, ge=1, le=12)
+
+    # Fields for dynamic rates
+    rate_datetime: Optional[datetime] = None
+    dst: Optional[bool] = None
```
Comment (ClemAtt) [inline opentariff/models/models.py:143]: Can we remove `dst`

**Review by ClemAtt — APPROVED** [review]
